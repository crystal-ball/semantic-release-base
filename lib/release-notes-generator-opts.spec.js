'use strict'

const { Readable, Writable } = require('stream')
/* eslint-disable import/no-extraneous-dependencies */
const conventionalChangelogWriter = require('conventional-changelog-writer')

const { writerOpts } = require('./release-notes-generator-opts')

/** Generate a context with overrides */
const gnContext = (overrides) => ({
  version: '1.3.0',
  currentTag: 'v1.3.0',
  previousTag: 'v1.2.0',
  host: 'https://github.com',
  owner: 'crystal-ball',
  repository: 'semantic-release-base',
  commit: 'commit',
  issue: 'issues',
  linkCompare: true,
  linkReferences: true,
  isPatch: false,
  date: '2019-05-26', // Normally autogenerated but passed here for snapshots
  ...overrides,
})

/** Generate a commit with overrides */
const gnCommit = ({
  tag = 'New',
  message = 'Hecka rad test message',
  notes = [],
  references = [],
} = {}) => ({
  header: `${tag}: ${message}`,
  tag,
  message,
  hash: '8d14946c50464dcf99cf5a9bd701262b2e3a2b4a',
  notes,
  references,
})

/**
 * Utility fn will create and pipe streams for creating a set of release notes
 * with the conventional-changelog-writer and this package's configs
 */
const generateReleaseNotes = (context, commits) =>
  new Promise((resolve) => {
    let releaseNotes = ''
    class MyWritable extends Writable {
      // eslint-disable-next-line class-methods-use-this, no-underscore-dangle
      _write(chunk, _, callback) {
        releaseNotes += chunk.toString()
        callback()
      }
    }

    const commitsStream = new Readable()
    commits.forEach((commit) => {
      commitsStream.push(JSON.stringify(commit))
    })
    commitsStream.push(null)

    commitsStream
      .pipe(conventionalChangelogWriter(context, writerOpts))
      .pipe(new MyWritable())

    commitsStream.on('end', () => {
      resolve(releaseNotes)
    })
  })

describe('Changelog writer configs', () => {
  test('All release notes features are rendered correctly', async () => {
    // Create a single release that should have an example of every major feature
    const releaseNotes = await generateReleaseNotes(gnContext(), [
      gnCommit({
        tag: 'Fix',
        message: 'Fixing broken assertion in test',
        references: [{ issue: '75' }],
      }),
      gnCommit({
        tag: 'Fix',
        message: 'Fixing broken assertion in test',
        references: [{ owner: 'cool-org', repository: 'cool-package', issue: '75' }],
      }),
      gnCommit({
        tag: 'Fix',
        message: 'Fixed a lot of broken things in this release',
        references: [{ issue: '100' }],
      }),
      gnCommit({ tag: 'Docs', message: 'Added docs on config overrides' }),
      gnCommit({ tag: 'Docs', message: 'Add Roadmap tasks for release notes' }),
      gnCommit({ tag: 'Build', message: 'Update Github Actions config' }),
      gnCommit({ tag: 'Upgrade', message: 'ESLint to latest' }),
      gnCommit({
        tag: 'New',
        message: 'Added a new feature that is rad',
        references: [{ issue: '80' }],
      }),
      gnCommit({
        tag: 'Update',
        message: 'Added additional overrides to config generator.',
        references: [{ issue: '99' }],
      }),
    ])
    expect(releaseNotes).toMatchSnapshot()
  })
})
