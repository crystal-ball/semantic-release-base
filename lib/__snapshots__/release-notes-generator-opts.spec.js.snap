// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Changelog writer configs All release notes features are rendered correctly 1`] = `
"## [1.3.0](https://github.com/crystal-ball/semantic-release-base/compare/v1.2.0...v1.3.0) (2019-05-26)


#### ðŸ’– New

* Added a new feature that is rad([8d14946](https://github.com/crystal-ball/semantic-release-base/commit/8d14946c50464dcf99cf5a9bd701262b2e3a2b4a)), closes [#80](https://github.com/crystal-ball/semantic-release-base/issue/80)

#### âœ¨ Update

* Added additional overrides to config generator.([8d14946](https://github.com/crystal-ball/semantic-release-base/commit/8d14946c50464dcf99cf5a9bd701262b2e3a2b4a)), closes [#99](https://github.com/crystal-ball/semantic-release-base/issue/99)

#### ðŸ›  Fix

* Fixed a lot of broken things in this release([8d14946](https://github.com/crystal-ball/semantic-release-base/commit/8d14946c50464dcf99cf5a9bd701262b2e3a2b4a)), closes [#100](https://github.com/crystal-ball/semantic-release-base/issue/100)
* Fixing broken assertion in test([8d14946](https://github.com/crystal-ball/semantic-release-base/commit/8d14946c50464dcf99cf5a9bd701262b2e3a2b4a)), closes [#75](https://github.com/crystal-ball/semantic-release-base/issue/75)
* Fixing broken assertion in test([8d14946](https://github.com/crystal-ball/semantic-release-base/commit/8d14946c50464dcf99cf5a9bd701262b2e3a2b4a)), closes [cool-org/cool-package#75](https://github.com/cool-org/cool-package/issue/75)

#### Docs, Upgrade, Build

* Add Roadmap tasks for release notes([8d14946](https://github.com/crystal-ball/semantic-release-base/commit/8d14946c50464dcf99cf5a9bd701262b2e3a2b4a))
* Added docs on config overrides([8d14946](https://github.com/crystal-ball/semantic-release-base/commit/8d14946c50464dcf99cf5a9bd701262b2e3a2b4a))
* ESLint to latest([8d14946](https://github.com/crystal-ball/semantic-release-base/commit/8d14946c50464dcf99cf5a9bd701262b2e3a2b4a))
* Update Github Actions config([8d14946](https://github.com/crystal-ball/semantic-release-base/commit/8d14946c50464dcf99cf5a9bd701262b2e3a2b4a))
"
`;
